app/src/control_msg.c

```diff
@@ -93,6 +128,74 @@ control_msg_serialize(const struct control_msg *msg, unsigned char *buf) {
     }
 }
 
+void
+control_msg_log(const struct control_msg *msg) {
+    switch (msg->type) {
+        case CONTROL_MSG_TYPE_INJECT_KEYCODE:
+            LOGV("Input: key %-4s code=%d repeat=%" PRIu32 " meta=%06lx",
+                 ENUM_TO_DESCRIPTION(android_keyevent_action_descriptions,
+                                     msg->inject_keycode.action),
+                 (int) msg->inject_keycode.keycode,
+                 msg->inject_keycode.repeat,
+                 (long) msg->inject_keycode.metastate);
+            break;
+        case CONTROL_MSG_TYPE_INJECT_TEXT:
+            LOGV("Input: text \"%s\"", msg->inject_text.text);
+            break;
+        case CONTROL_MSG_TYPE_INJECT_TOUCH_EVENT: {
+            size_t action = msg->inject_touch_event.action & AMOTION_EVENT_ACTION_MASK;
+            LOGV("Input: touch %-4s position=%" PRIi32 "x%" PRIi32 " pressure=%g buttons=%06lx",
```

Comment:
I omitted `pointer_id` from this message because it didn't didn't seem very useful.

Reply:
I think you should log it, it allows to distinguish several simultaneous fingers on touch events.